### DISCUSSION TITLE
How do you manage buffers in LazyVim?

### DISCUSSION THREAD

**Discussion by pprotas** [LazyVim/LazyVim#218] [General]
I find myself opening a bunch of buffers and the tabline at the top gets very full. At a certain point I lose track of which buffers I am working on.

I try to use the `<leader>,` shortcut to see open buffers, use `H` and `L` to switch between buffers in the tabline and also use the `<leader>e` followed by `<` to see all open buffers in a treeview. But one thing that annoys me: there are TOO MANY open buffers. Usually I only work on 2-3 files at a time, and all other buffers are only there for reference purposes.

How do you manage your buffers while programming? I think if I had a way to "pin" the most important buffers it would be perfect. Add on to that a way to close ALL buffers except the ones I have pinned.

**Comment by dpetka2001** [comment]
You could try [`harpoon`](https://github.com/ThePrimeagen/harpoon) to pin your "important" buffers and use other ways for the rest of the buffers. I personally like `:Telescope buffers`.

**Comment by briandipalma** [comment]
Delete the buffers you don't need any more using `<leader>bd`. And switch buffers using `<leader>fb`?

**Comment by folke** [comment]
Additionally, to the previous answers, https://github.com/akinsho/bufferline.nvim also supports pinning tabs. I don't use that functionality myself, but worth looking at it.

Personally, my workfow is:
* I use `H` and `L` if the buffer I want to go to is visually close to where I'm at
* Otherwise, if I know the buffer is open, I use `<leader>,`
* For any other file, I use `<leader><space>`

I also close buffers I know I just needed to quickly look something up.

And I of course use `<c-o>` and `<c-i>`, with `gd` all the time to navigate the code.

Also `<leader>ss` to quickly jump to a function in the buffer I'm on.

**Reply by surmish** [reply]
Was there are particular reason why `[b` and `]b` were added for the same functionality?

**Reply by briandipalma** [reply]
They aren't the same, one is previous the other is next buffer.

**Reply by surmish** [reply]
I meant this from `keymaps.lua`

```lua
if Util.has("bufferline.nvim") then
  map("n", "<S-h>", "<cmd>BufferLineCyclePrev<cr>", { desc = "Prev buffer" })
  map("n", "<S-l>", "<cmd>BufferLineCycleNext<cr>", { desc = "Next buffer" })
  map("n", "[b", "<cmd>BufferLineCyclePrev<cr>", { desc = "Prev buffer" })
  map("n", "]b", "<cmd>BufferLineCycleNext<cr>", { desc = "Next buffer" })
else
```

No need for duplicate mappings, right?

**Reply by briandipalma** [reply]
There are many duplicate mappings like that, I think they make sense. `[` and `]` are traditional prev/next mappings in vim and some people may have built up muscle memory using them.

**Comment by dpetka2001** [comment]
In telescope configuration I also have
```lua
    opts = {
      defaults = {
        mappings = {
          i = {
            ["<C-b>"] = function(...)
              return require("telescope.actions").delete_buffer(...)
            end,
          },
          n = {
            ["<C-b>"] = function(...)
              return require("telescope.actions").delete_buffer(...)
            end,
          },
        },
      },
    },
```
so that I can quickly delete buffers from Telescope pickers with `C-b` without the need to go to every single buffer and then `<leader>bd` to delete it.

**Reply by briandipalma** [reply]
Thanks, this is a great keymap. I disabled bufferline so this lets me select all the buffers I don't want open and close them in one operation. I think this should be in LazyVim by default. I had to use `alt-d` though as `c-b` and `c-x` clashed with other keymaps.

**Comment by briandipalma** [comment]
Is it possible to add a `Tips` section next to the `Recipes` section in the documentation? The workflow description above is useful and worth making it documentation IMO. I'm tempted to copy the above telescope configuration and the lazygit full screen code from the other discussion into the `Recipes` section it's just I'm worried it will end up getting packed with stuff...

**Reply by folke** [reply]
I just added a Tips section to the website (will be live in a bit).

Feel free to submit a PR!
See https://github.com/LazyVim/lazyvim.github.io/blob/main/docs/configuration/tips.md

**Comment by pprotas** [comment]
Great tips! I added @dpetka2001's Telescope mappings. 

If anyone else wants to use pins, here is an example config I just made up:

```lua
return {
  {
    "akinsho/bufferline.nvim",
    keys = {
      {
        "<leader>bp",
        "<Cmd>BufferLineTogglePin<CR>",
        desc = "Toggle pin",
      },
      {
        "<leader>bP",
        "<Cmd>BufferLineGroupClose ungrouped<CR>",
        desc = "Delete non-pinned buffers",
      },
    },
  },
}

```

You can use `<leader>bp` to pin your important tabs and use the other methods described in the thread to switch between them. If you are annoyed at the amount of tabs, you can use `<leader>bP` to get rid of all tabs that aren't pinned!

**Reply by folke** [reply]
This is actually really neat! And perfect keymaps.

I just added this to LazyVim :)

https://github.com/LazyVim/LazyVim/commit/3af91550be67d9699567727fffcae8aabafc0898

**Comment by MatejBransky** [comment]
I don't use bufferline because it just distracts me but I've found helpful this Telescope setting `sort_mru` which sorts buffers based on most recent usage so now I can just use `<leader>,` and use h,j,k,l to switch between the most desired ones. 
```
      pickers = {
        -- Sorts all buffers after most recent used.
        buffers = {
          sort_mru = true,
        },
      }
```

**Reply by MatejBransky** [reply]
Another thing that helped me a lot - VS Code-like fuzzy searching algorithm [`natecraddock/telescope-zf-native.nvim`](https://github.com/natecraddock/telescope-zf-native.nvim) so searching between opened buffers is really fast if you type e.g. only upper cases. And for frequently used buffers I use Harpoon as @dpetka2001 has already mentioned too.

**Reply by dpetka2001** [reply]
Didn't know about that. Personally, I use `nvim-telescope/telescope-fzf-native.nvim`. I read on the description of the plugin you mentioned, but could you also, if possible, elaborate a bit more about comparisons between the 2? I would really appreciate some of your personal thoughts based on your experience.

**Reply by MatejBransky** [reply]
I was used to VS Code searching algorithm which prioritizes via smart case and searched term in filenames:
![image](https://github.com/LazyVim/LazyVim/assets/20944757/be4b7af4-8124-49b2-aaa7-6766be792390)

But neither of Telescope recommended solutions achieves the same experience:
![image](https://github.com/LazyVim/LazyVim/assets/20944757/32ee1688-fc0c-4d72-875e-4437c0a34430)

This was one of the main bottlenecks which I had in the neovim because I was terribly slow at searching files across the whole monorepo. Finally I've found the mentioned algorithm which consistently gives good results:
<img width="1029" alt="image" src="https://github.com/LazyVim/LazyVim/assets/20944757/e1f7015e-f1f0-4c5e-b00a-31173cd86d85">

**Reply by dpetka2001** [reply]
I see. Appreciate your feedback. Will try it out :smile:

**Comment by sp71** [comment]
Any thoughts on adding harpoon to lazy Vim base config?

**Reply by dpetka2001** [reply]
You can always add it to your personal spec. I don't think it's something that everyone would use. I added it to my spec on my own.
